compiler-rt/lib/hwasan/hwasan_checks.h

```diff
@@ -140,7 +140,7 @@ __attribute__((always_inline, nodebug)) static inline uptr ShortTagSize(
 
 __attribute__((always_inline, nodebug)) static inline bool
 PossiblyShortTagMatches(tag_t mem_tag, uptr ptr, uptr sz) {
-  DCHECK(IsAligned(ptr, kShadowAlignment));
+  DCHECK(IsAligned(ptr, sz));
```

Comment:
This change conflicts with CheckAddressSized @ line 186, where tail_sz can be anything

However can I see how this DCHECK fails for CheckAddress.

PossiblyShortTagMatches @ line 
```
return *(u8 *)(ptr | (kShadowAlignment - 1)) == ptr_tag;
```
expects that ptr points to the first bite of the granule.

Reply:
I suppose this one should guaranty that we can't get here with wrong alignment

```
IRBuilder<> IRB(O.getInsn());
  if (!O.TypeStoreSize.isScalable() && isPowerOf2_64(O.TypeStoreSize) &&
      (O.TypeStoreSize / 8 <= (1ULL << (kNumberOfAccessSizes - 1))) &&
      (!O.Alignment || *O.Alignment >= Mapping.getObjectAlignment() ||
       *O.Alignment >= O.TypeStoreSize / 8)) {
```

But I am not sure how :)

Reply:
I re-read your description, good point about `| (kShadowAlignment - 1)`

`DCHECK(IsAligned(ptr, sz));` is still not correct for CheckAddressSized. I guess we need to remove it.

Reply:
Thanks for looking into this, I am very unfamiliar with the hwasan code and was just trying to get the tests to pass. The always zero condition made me think this fix is correct and it would make sense for all accesses to be naturally aligned but I don't yet understand the short tag logic.

Reply:
No, that is not correct. It is actually `| (kShadowAlignment - 1)` not `&`, to access the last byte of the last granule (to get the short granule tag).

Reply:
yes, this means we don't care about pointer alignment in this function and we can remove DCHECK?

Reply:
Sorry, I got confused. I see no reason that we need this `DCHECK`, and yes it's clearly not legal for `hwasan_load1`, let's drop it?